src/library/dao/trackdao.cpp

```diff
@@ -1545,143 +1548,162 @@ bool TrackDAO::detectMovedTracks(QSet<TrackId>* pTracksMovedSetOld,
-        DbId oldTrackLocationId(deletedTrackQuery.value(idColumn));
-        filename = deletedTrackQuery.value(filenameColumn).toString();
-        duration = deletedTrackQuery.value(durationColumn).toInt();
+        QString oldTrackLocation = oldTrackQuery.value(oldLocationColumn).toString();
+        QString filename = oldTrackQuery.value(filenameColumn).toString();
+        // rather use duration then filesize as an indicator of changes. The filesize
+        // can change by adding more ID3v2 tags
+        const int duration = oldTrackQuery.value(durationColumn).toInt();
 
-        qDebug() << "TrackDAO::detectMovedTracks looking for a successor of" << filename << duration;
+        kLogger.info()
+                << "Looking for substitute of missing track location"
+                << oldTrackLocation;
 
         newTrackQuery.bindValue(":filename", filename);
         newTrackQuery.bindValue(":duration", duration);
-        if (!newTrackQuery.exec()) {
-            // Should not happen!
+        VERIFY_OR_DEBUG_ASSERT(newTrackQuery.exec()) {
             LOG_FAILED_QUERY(newTrackQuery);
+            continue;
         }
-        // WTF duplicate tracks?
-        if (newTrackQuery.size() > 1) {
-            LOG_FAILED_QUERY(newTrackQuery) << "Result size was greater than 1.";
-        }
-
-        const int query2idColumn = newTrackQuery.record().indexOf("id");
+        int newTrackLocationCount = 0;
+        TrackId newTrackId;
+        DbId newTrackLocationId;
+        QString newTrackLocation;
         while (newTrackQuery.next()) {
-            newTrackLocationId = DbId(newTrackQuery.value(query2idColumn));
+            newTrackLocation = newTrackQuery.value(
+                    newTrackQuery.record().indexOf("location")).toString();
+            VERIFY_OR_DEBUG_ASSERT(newTrackLocation != oldTrackLocation) {
+                continue;
+            }
+            kLogger.info()
+                    << "Found potential moved track location:"
+                    << newTrackLocation;
+            ++newTrackLocationCount;
+            newTrackId = TrackId(newTrackQuery.value(
+                    newTrackQuery.record().indexOf("track_id")));
+            newTrackLocationId = DbId(newTrackQuery.value(
+                    newTrackQuery.record().indexOf("location_id")));
+            DEBUG_ASSERT(newTrackLocationId.isValid());
+        }
+        switch (newTrackLocationCount) {
+        case 0:
+            kLogger.info()
+                    << "Found no substitute for missing track location"
+                    << oldTrackLocation;
+            continue;
+        case 1:
+            DEBUG_ASSERT(newTrackId.isValid());
+            DEBUG_ASSERT(newTrackLocationId.isValid());
+            kLogger.info()
+                    << "Found moved track location:"
+                    << oldTrackLocation
+                    << "->"
+                    << newTrackLocation;
+            break;
+        default:
+            kLogger.warning()
```

Comment:
This is the case we have tree duplicates and one disappears, or if we have duplicates and one is moved. 
It would be nice to detect the move. In this case use the just added track.

Reply:
I don't understand this comment. All the new have just been added. If multiple of them are candidates for one missing track, we cannot make a decision.

Reply:
Is this a possible case: 

* User has the same track in two folders. 
* He moved to a new library location.
* He expects that the tracks are still OK 

What will happen? I have not tested it yet. 

I assume currently both tracks remain missing after the move and pointing to the old location. I think it is better to assign both duplicates to the first of the two moved tracks than break a playlist.